jolt-core/src/subprotocols/grand_product.rs

```diff
@@ -1,530 +1,1699 @@
-            let claims_prod_right = &self.proof[i].claims_poly_B;
-            assert_eq!(claims_prod_left.len(), claims_prod_vec.len());
-            assert_eq!(claims_prod_right.len(), claims_prod_vec.len());
+/// Represents a single layer of a single grand product circuit using a sparse vector,
+/// i.e. a vector containing (index, value) pairs.
+/// Nodes with value 1 are omitted from the sparse vector.
+/// Like `DenseGrandProductLayer`, a `SparseGrandProductLayer` is assumed to be
+/// arranged in "interleaved" order:
+///      Λ        Λ        Λ        Λ
+///     / \      / \      / \      / \
+///   L0   1   L1   R1   1   1   L3   1   <- This is layer would be represented as [(0, L0), (2, L1), (3, R1), (6, L3)]
+pub type SparseGrandProductLayer<F> = Vec<(usize, F)>;
+
+/// A "dynamic density" grand product layer can switch from sparse representation
+/// to dense representation once it's no longer sparse (after binding).
+#[derive(Debug, Clone, PartialEq)]
+pub enum DynamicDensityGrandProductLayer<F: JoltField> {
+    Sparse(SparseGrandProductLayer<F>),
+    Dense(DenseGrandProductLayer<F>),
+}
 
-            for i in 0..claims_prod_vec.len() {
-                transcript.append_scalar(b"claim_prod_left", &claims_prod_left[i]);
-                transcript.append_scalar(b"claim_prod_right", &claims_prod_right[i]);
+/// The threshold at which a `DynamicDensityGrandProductLayer` switches from sparse to
+/// dense representation. If the layer has >DENSIFICATION_THRESHOLD fraction of non-1
+/// values, it'll switch to the dense representation.
+const DENSIFICATION_THRESHOLD: f64 = 0.8;
+
+impl<F: JoltField> DynamicDensityGrandProductLayer<F> {
+    /// Computes the grand product layer that is output by this layer.
+    ///     L0'      R0'      L1'      R1'     <- output layer
+    ///      Λ        Λ        Λ        Λ
+    ///     / \      / \      / \      / \
+    ///   L0   R0  L1   R1  L2   R2  L3   R3   <- this layer
+    ///
+    /// If the current layer is dense, the output layer will be dense.
+    /// If the current layer is sparse, but already not very sparse (as parametrized by
+    /// `DENSIFICATION_THRESHOLD`), the output layer will be dense.
+    /// Otherwise, the output layer will be sparse.
+    pub fn layer_output(&self, output_len: usize) -> Self {
+        match self {
+            DynamicDensityGrandProductLayer::Sparse(sparse_layer) => {
+                #[cfg(test)]
+                let product: F = sparse_layer.iter().map(|(_, value)| value).product();
+
+                if (sparse_layer.len() as f64 / (output_len * 2) as f64) > DENSIFICATION_THRESHOLD {
+                    // Current layer is already not very sparse, so make the next layer dense
+                    let mut output_layer: DenseGrandProductLayer<F> = vec![F::one(); output_len];
+                    let mut next_index_to_process = 0usize;
+                    for (j, (index, value)) in sparse_layer.iter().enumerate() {
```

Comment:
Can be parallelized fairly easily in the "sparse to dense" case and less easily in the "sparse to sparse" case. Looks like only 1% of e2e runtime on 16 cores, but probably more significant on a 64 core machine.

Reply:
yeah, we parallelizing these sparse layers is tricky, but may be worth revisiting so that we scale better with # cores